libsignal-service/src/gv2/api.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    collections::HashMap,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use crate::{
+    configuration::Endpoint,
+    prelude::{PushService, ServiceError},
+};
+
+use rand::RngCore;
+use uuid::Uuid;
+use zkgroup::{
+    auth::AuthCredentialResponse, groups::GroupSecretParams, ServerPublicParams,
+};
+
+use super::models::CredentialResponse;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CredentialsCacheError {
+    #[error("failed to read values from cache: {0}")]
+    ReadError(String),
+    #[error("failed to write values from cache: {0}")]
+    WriteError(String),
+}
+
+pub trait CredentialsCache {
+    fn clear(&self) -> Result<(), CredentialsCacheError>;
+    fn read(
+        &self,
+    ) -> Result<HashMap<i64, AuthCredentialResponse>, CredentialsCacheError>;
+    fn write(
+        &self,
+        value: &HashMap<i64, AuthCredentialResponse>,
+    ) -> Result<(), CredentialsCacheError>;
+}
```

Comment:
I think I'd rather see a `read(i16) -> Result<ACR>; write(i64, ACR)` for this, possibly with a read_many and write_many if really necessary.

Reply:
The read/write_many can have a default-impl, and should probably take some iterator forms...

Reply:
We can also supply an in-memory and in-xdg-cache implementation for these!

Reply:
so... it doesn't really help to expose the internal hash map because it works like this:
1. Get 7 days of credentials
2. Use the cache
3. At some point, you don't have a new day in the cache, trash everything, and write the next 7 days

Since the current back-end I have is `sled`, it does make sense to have key-value in the interface, but it's not really useful as far as the actual feature is concerned.

Reply:
Is this API per-group, or per-client? It's just weird to have a `read()` without parameters (except for self)

Reply:
It's per-client, so let's say "global".

Reply:
And the cached data is for all groups together then? I'd think there's a cached authorization proof per group.

Reply:
You're absolutely right, I got mixed up in the onion layers.